### PR TITLE
Implement Tree.update and Tree.update_tree

### PR DESCRIPTION
**Pull request by CraigFe**
Resolves https://github.com/mirage/irmin/issues/1270.

`Tree.update_tree` is quite complicated, with many edge cases that I've tried to test as much as is practical. Fortunately, the functions `Tree.add`, `Tree.add_tree`, `Tree.remove` and `Tree.update` can all be implemented in terms of it with little to no performance cost, which I've done here.

### CODE REVIEW

**Review by Ngoguey42 — COMMENTED** [review]
I checked your implementation in detail, it looks sound to me.

The performances are identical to master regarding bench/tree.exe's trace mode.

File: src/irmin/tree.ml [hunk 1]
```diff
@@ -1483,7 +1488,9 @@ module Make (P : Private.S) = struct
       | `Node n ->
           let* m = Node.to_map n in
           let bindings = m |> get_ok |> StepMap.bindings in
-          (node [@tailcall]) [] bindings (fun n -> k (`Tree n))
+          (node [@tailcall]) [] bindings (fun n ->
+              let n = List.sort (fun (s, _) (s', _) -> compare_step s s') n in
```
[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin/tree.ml:1492]: Yes: we rely on this ordering implicitly in tests, since equality / pretty-printing is derived automatically for concrete trees from the type. We could change the generic operations to sort the bindings first, but it's simplest just to have a well-defined ordering when going from regular trees.

File: src/irmin/tree.ml [hunk 2]
```diff
@@ -1082,113 +1075,125 @@ module Make (P : Private.S) = struct
...
+              let* old_binding = Node.findv parent_node file in
+              let new_binding = f old_binding in
+              match (old_binding, new_binding) with
+              | None, None -> k Unchanged
+              | None, Some (`Contents _ as t) -> with_new_child t
+              | None, Some (`Node n as t) -> (
+                  Node.is_empty n >>= function
+                  | Ok true ->
+                      (* Prune empty node introduced by the user. *)
+                      k Unchanged
```
[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin/tree.ml:1132]: Checking for emptiness may require an extra read from the repository, so is worth avoiding in the redundant cases.

In fact, in hindsight I think it's probably not worth doing this at all. I had thought that the previous version of `Tree.add_tree` was performing this check, but it doesn't seem to be. We can probably just be consistent with that here and simplify the implementation quite a bit; have pushed a commit that does that.

File: src/irmin/tree.ml [hunk 3]
```diff
@@ -1082,113 +1075,125 @@ module Make (P : Private.S) = struct
...
+                  | Ok false | Error (`Dangling_hash _) -> with_new_child t)
+              | Some _, None -> Node.remove parent_node file >>= changed
+              | Some old_value, Some (`Node n as t) -> (
+                  Node.is_empty n >>= function
+                  | Ok true ->
+                      (* Prune empty node introduced by the user. *)
+                      Node.remove parent_node file >>= changed
```
[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin/tree.ml:1139]: Likewise, this doesn't occur after simplifying the semantics.

File: src/irmin/tree.ml [hunk 4]
```diff
@@ -1082,113 +1075,125 @@ module Make (P : Private.S) = struct
...
+                  | Ok false | Error (`Dangling_hash _) -> (
+                      match old_value with
+                      | `Contents _ -> with_new_child t
+                      | `Node old -> (
+                          match Node.maybe_equal old n with
+                          | `True -> k Unchanged
+                          | `Maybe | `False -> with_new_child t)))
```
[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin/tree.ml:1146]: With the simplified behaviour, these two cases are symmetric anyway.

File: src/irmin/tree_intf.ml [hunk 5]
```diff
@@ -117,6 +117,15 @@ module type S = sig
   (** [add t k c] is the tree where the key [k] is bound to the contents [c] but
       is similar to [t] for other bindings. *)
 
+  val update :
+    t ->
+    key ->
+    ?metadata:metadata ->
+    (contents option -> contents option) ->
```
[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin/tree_intf.ml:124]: This is already implied by the doc-string of `find`, but I've added an extra sentence to make it clearer.

[the remark this answers is not in the document]
Comment (CraigFe) [reply src/irmin/tree_intf.ml:124]: I think this is already implied by the doc-string of `find`, but I've added an extra sentence to make it clearer.

**Review by Ngoguey42 — APPROVED** [review]

File: src/irmin/tree.ml [hunk 6]
```diff
@@ -1082,113 +1076,120 @@ module Make (P : Private.S) = struct
...
+          | true -> root_tree
+          | false -> `Node empty_node
+        in
+        match f (Some root_tree) with
+        (* Here we consider "deleting" a root contents value or node to consist
+           of changing it to an empty node. Note that this introduces
+           sensitivity to ordering of subtree operations: updating in a subtree
+           and adding the subtree are not necessarily commutative. *)
+        | None -> Lwt.return empty_tree
+        | Some (`Node _ as t) -> Lwt.return t
```
Comment (icristescu) [inline src/irmin/tree.ml:1113]: Does it make sense to check for equality between old and new here too?

Comment (CraigFe) [reply src/irmin/tree.ml]: It does, thanks!

File: src/irmin-test/store.ml [hunk 7]
```diff
@@ -1257,24 +1258,34 @@ module Make (S : S) = struct
...
+        >>= with_binding [ "foo"; "b"; "c" ] "2"
+        >>= with_binding [ "bar"; "d" ] "3"
+        >>= with_binding [ "e" ] "4"
+      in
+      let* t0 = c0 |> S.Tree.to_concrete >|= S.Tree.of_concrete in
+      let* () =
+        let+ d0 = S.Tree.diff c0 t0 in
+        check_diffs "concrete roundtrip" [] d0
+      in
+      let () = Format.printf "Boo: %a@.\n" (T.pp_dump S.tree_t) c0 in
```
Comment (icristescu) [inline src/irmin-test/store.ml:1273]: Looks like you forgot this printf here

**Review by icristescu — APPROVED** [review]
Thanks!

File: src/irmin-test/store.ml (hunk 7 above)
Comment (CraigFe) [reply src/irmin-test/store.ml:1273]: Indeed, thanks :stuck_out_tongue:

**Comment by CraigFe** [comment]
Will merge once the CI passes.
